### DISCUSSION TITLE
Adding Branch Protection Rules through endpoint API

### DISCUSSION THREAD

**Discussion by seanmalloy-crest** [community/community#24762] [Apps, API and Webhooks]
<p>I am trying to add a <a href="https://docs.github.com/en/rest/reference/repos#update-pull-request-review-protection" rel="noopener nofollow ugc">Pull request review protection</a> rule to one of the repositories I am working on, and using the following API call to add it:</p>
<pre><code class="lang-auto">curl -X PATCH -H "Accept: application/vnd.github.v3+json" https://api.github.com/repos/$REPOORG/$REPO/branches/main/protection/required_pull_request_reviews -d '{"require_code_owner_reviews":true}'
</code></pre>
<p>I keep getting this error message. When I try to check the repository branch, I get a 404 for found error message: <a href="https://api.github.com/repos/seanmalloy-crest/test_environment/branches/main/protection" rel="noopener nofollow ugc">https://api.github.com/repos/seanmalloy-crest/test_environment/branches/main/protection</a>.<br>
i have admin permission to change the repository, but I was wondering what API query I should use to add the branch protection rule with. I have tried using the following API call, but that didn’t seem to work either:</p>
<pre><code class="lang-auto">hub api -X PUT -H "Accept: application/vnd.github.v3+json" /repos/$REPOORG/$REPO/branches/main/protection --raw-field '{"dismissal_restrictions": {},"dismiss_stale_reviews": false,"require_code_owner_reviews": true,"required_approving_review_count": 1}'
</code></pre>
<p>And I also tried this command and got this error message instead:</p>
<pre><code class="lang-auto">hub api -t -X PUT -H 'Accept: application/vnd.github.v3+json' /repos/splunk/splunk-add-on-for-ucc-example/branches/main/protection/required_pull_request_reviews --field require_code_owner_reviews=true
.message	Branch not protected
.documentation_url	https://docs.github.com/rest/reference/repos#update-pull-request-review-protection
</code></pre>
<p>I would really appreciate the help!</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/2/2/22dd80991c30fdf5b476fe96128fdfd31081d56e.png" data-download-href="/uploads/short-url/4YqNDmlYR4630e8U0OhNElApesm.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/60528866/181086607-d48e2c3f-be16-485c-9a61-cc882ea26e5a.png" alt="image" data-base62-sha1="4YqNDmlYR4630e8U0OhNElApesm" width="690" height="177" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/2/2/22dd80991c30fdf5b476fe96128fdfd31081d56e_2_690x177.png, https://aws1.discourse-cdn.com/github/optimized/3X/2/2/22dd80991c30fdf5b476fe96128fdfd31081d56e_2_1035x265.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/2/2/22dd80991c30fdf5b476fe96128fdfd31081d56e_2_1380x354.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/2/2/22dd80991c30fdf5b476fe96128fdfd31081d56e_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1822×468 75.3 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by seanmalloy-crest** [comment]
<p>Looking at other posts, I am getting the same issue asked about here:</p><aside class="quote quote-modified" data-post="1" data-topic="14612">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/user-cannot-access-branch-protection-endpoint-via-api/14612">User cannot access branch protection endpoint via API</a> <a class="badge-wrapper  bullet" href="/c/github-ecosystem/37"><span class="badge-category-bg" style="background-color: #28a745;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Discussion and support using GitHub’s REST, and GraphQL APIs, building Applications and OAuth Apps, the GitHub Marketplace, and third party integrations.">GitHub Ecosystem</span></a>
  </div>
  <blockquote>
    We ran into a problem that is quite confusing - we have a user trying to access the endpoint: 
/repos/:owner/:repo/branches/:branch/protection 
via a curl call to the API. He consistently receives a “404 Not found 
#get-required-status-checks-of-protected-branch” type error. 
His access token is valid and works across other endpoints in the same repo, so it’s not a credentials issue, for example. He *can* access: 
/repos/:owner/:repo/branches/:branch 
and it even lists the ‘protection’ url it cl…
  </blockquote>
</aside>

<p>I personally do not have the required access though I am using a Github Token that should have admin permission to that repository.</p>

**Comment by nethgato** [comment, marked as answer]
<p>Hey <a class="mention" href="/seanmalloy-crest">@seanmalloy-crest</a> how are you going, aside from these API troubles? Well, I hope!</p>
<p>You mentioned:</p>
<blockquote>
<p>I am using a Github Token that should have admin permission to that repository.</p>
</blockquote>
<p>But in your example curl, there was no token being applied to your calls, that I could see. At least, not from your screenshot above. Our REST Authentication docs are here:</p>
<aside class="onebox githubfolder">
  <header class="source">
      <img src="https://docs.github.com/assets/images/site/favicon.svg" class="site-icon" width="16" height="16">
      <a href="https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api#authentication" target="_blank" rel="noopener">docs.github.com</a>
  </header>
  <article class="onebox-body">
    

<h3><a href="https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api#authentication" target="_blank" rel="noopener">Getting started with the REST API - GitHub Docs</a></h3>

<p><a href="https://docs.github.com/en/rest/guides/getting-started-with-the-rest-api#authentication" target="_blank" rel="noopener">//docs.github.com/en/rest/guides/getting-started-with-the-rest-api</a></p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

<p>Because a 404 not found could easily mean permissions issues and if you aren’t passing a token with valid permission, would explain what you see.</p>
<p>Let us know if you continue to have trouble after properly authenticating your calls! :bow:</p>

**Comment by seanmalloy-crest** [comment]
<p>Thanks, that seemed to help a lot!</p>
<p>Here is the final curl call I made, in case if anyone else has a similar issue:</p>
<pre><code class="lang-auto">curl -X PUT -u $GITHUB_USER:$GITHUB_TOKEN -H "Accept: application/vnd.github.luke-cage-preview+json" https://api.github.com/repos/$REPOORG/$REPO/branches/main/protection -d '{"required_status_checks": null,"enforce_admins": null,"required_pull_request_reviews" : {"dismissal_restrictions": {},"dismiss_stale_reviews": false,"require_code_owner_reviews": true,"required_approving_review_count": 1},"restrictions":null}'
</code></pre>
